url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/parser/ExprLexer.java

```diff
@@ -0,0 +1,266 @@
+        case '*':
+            lastChar = 0;
+            return ExprToken.MULTIPLY;
+        case '/':
+            lastChar = 0;
+            return ExprToken.DIVIDE;
+        case ',':
+            lastChar = 0;
+            return ExprToken.COMMA;
+        case '&':
+            lastChar = 0;
+            return ExprToken.STRING_CONCAT;
+        case '<':
+        case '>':
+        case '=':
+            return readComparisonOperator();
+        case '\'':
+            return readQuotedVariable();
+        case '{':
+            lastChar = 0;
+            return ExprToken.OPEN_BRACE;
+        case '}':
+            lastChar = 0;
+            return ExprToken.CLOSE_BRACE;
+        case ';':
+            lastChar = 0;
+            return ExprToken.SEMI_COLON;
+        case '^':
+            lastChar = 0;
+            return ExprToken.POWER;
+        case -1:
+        case 0xffff:
+            return null;
+        }
+
+        if (!Character.isJavaIdentifierStart(lastChar)) {
+            throw new IOException("Invalid token found: " + lastChar);
+        }
+
+        return readVariableOrFunction();
+    }
+
+    private ExprToken readQuotedVariable() throws IOException {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append('\'');
+        while (lastChar != -1) {
+            lastChar = reader.read();
+            if (lastChar == '\'') {
+                lastChar = reader.read();
+                if (lastChar == '\'') {
+                    sb.append('\'');
+                } else {
+                    break;
+                }
+            } else {
+                sb.append((char) lastChar);
+            }
+        }
+        sb.append('\'');
+
+        // Now read the rest of the variable
+        while (isVariablePart(lastChar)) {
+            sb.append((char) lastChar);
+            lastChar = reader.read();
+        }
+
+        return new ExprToken(ExprTokenType.Variable, sb.toString());
+    }
+
+    private ExprToken readVariableOrFunction() throws IOException {
+        StringBuilder sb = new StringBuilder();
+
+        while (isVariablePart(lastChar)) {
+            sb.append((char) lastChar);
+            lastChar = reader.read();
+        }
+
+        if (Character.isWhitespace(lastChar)) {
+            lastChar = reader.ignoreWhitespace();
+        }
+
+        if (lastChar == '(') {
+            lastChar = 0;
+            return new ExprToken(ExprTokenType.Function, sb.toString());
+        } else {
+            return new ExprToken(ExprTokenType.Variable, sb.toString());
+        }
+    }
+
+    private boolean isVariablePart(int lastChar) {
+        return Character.isJavaIdentifierPart(lastChar) || lastChar == '!' ||
+                lastChar == ':';
+    }
+
+    private ExprToken readString() throws IOException {
+        String str = unescapeString(reader);
+        lastChar = 0;
+        return new ExprToken(ExprTokenType.String, str);
+    }
+
+    private ExprToken readNumber() throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 175-186, lines 188-198 into separate methods.